server/automute.go

```diff
@@ -0,0 +1,214 @@
+package main
+
+import (
+	"database/sql"
+	"fmt"
+	"strconv"
+
+	"github.com/mattermost/mattermost/server/public/model"
+	"github.com/pkg/errors"
+)
+
+const (
+	NotifyPropAutomuted = "msteams_automute"
+
+	// This preference tracks if we've enabled automuting for the user by muting all of their connected channels.
+	PreferenceNameAutomuteEnabled = "automute_enabled"
+)
+
+func (p *Plugin) tryEnableAutomute(userID string, skipConnectedCheck, skipPrimaryPlatformCheck bool) (bool, error) {
+	enabled, err := p.shouldEnableAutomuteForUser(userID, skipConnectedCheck, skipPrimaryPlatformCheck)
+	if err != nil {
+		return false, err
+	} else if !enabled {
+		return false, nil
+	}
+
+	return p.setAutomuteEnabledForUser(userID, true)
+}
+
+// shouldEnableAutomuteForUser returns true if the given user is both connected to Teams and has their primary platform
+// set to Teams.
+//
+// skipConnectedCheck and skipPrimaryPlatformCheck can be used to skip the respective checks if the calling code has
+// already performed those checks.
+func (p *Plugin) shouldEnableAutomuteForUser(userID string, skipConnectedCheck, skipPrimaryPlatformCheck bool) (bool, error) {
+	var connected bool
+	if skipConnectedCheck {
+		connected = true
+	} else {
+		var err error
+		connected, err = p.isUserConnected(userID)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	var teamsPrimary bool
+	if skipPrimaryPlatformCheck {
+		teamsPrimary = true
+	} else {
+		teamsPrimary = p.isUsersPrimaryPlatformTeams(userID)
+	}
+
+	if !connected || !teamsPrimary {
+		return false, nil
```

Comment:
0/5 on the performance value of it (not sure how often will this function be called) but we can early return once we have the `connected` value.

Reply:
That's a good point about returning early. I think I forgot to do that because I was super focused on performance which is why `skipConnectedCheck` and `skipPrimaryPlatform` exist. I expected this method to be called whenever anything happened that might trigger automuting, but since I added the `automute_enabled` preference, most code can just check that instead which means performance is less of an issue here.

I'm going to return early when `!connected` like you suggested, but I might also remove the `skipX` parameters to simplify this logic as well. It'll make performance slightly worse overall, but I think that would be overoptimization overall